Tengo una gran afición la astronomía y me ha parecido genial la oportunidad de relacionar lo que es mi oficio con una de mis grandes pasiones.

El objetivo es conectarme a una API y obtener datos de ella.

Selecciono la API de la NASA, del Servicio de Detección de objetos Cercanos a la Tierra (NeoWs) que se encuentra dentro de la página https://api.nasa.gov/ para ver si puedo obtener algunos datos relacionados con objetos entre los días 10 y 12 de enero de 2025.

In [ ]:
import requests
import pandas as pd

# Hago la solicitud a  la API de la NASA según lo indican en su sitio web

URL = "https://api.nasa.gov/neo/rest/v1/feed?start_date=2025-01-10&end_date=2025-01-12&api_key=DEMO_KEY"
response = requests.get(URL)

if response.status_code == 200:
    data = response.json()
    print("Solicitud Exitosa")
    
else:
    print("Error en la solicitud, detalles:", response.text)

Solicitud Exitosa


Quiero saber como están estructurados los datos del objeto creado y los inspecciono mediante la función "keys"

In [2]:
data.keys()

dict_keys(['links', 'element_count', 'near_earth_objects'])

Dentro del campo near_earth_objects creo que es donde puedo encontrar los datos importantes que me interesan obtener, consulto la estructura para poder elegir los datos que me interesan para crear un dataframe con ellos.

El repositorio de Github me mostrará el resultado completo de una lista de 34 objetos, es una salida muy larga.

In [9]:
data['near_earth_objects']

{'2025-01-12': [{'links': {'self': 'http://api.nasa.gov/neo/rest/v1/neo/3313157?api_key=DEMO_KEY'},
   'id': '3313157',
   'neo_reference_id': '3313157',
   'name': '(2006 BA)',
   'nasa_jpl_url': 'https://ssd.jpl.nasa.gov/tools/sbdb_lookup.html#/?sstr=3313157',
   'absolute_magnitude_h': 27.6,
   'estimated_diameter': {'kilometers': {'estimated_diameter_min': 0.0080270317,
     'estimated_diameter_max': 0.0179489885},
    'meters': {'estimated_diameter_min': 8.0270316728,
     'estimated_diameter_max': 17.948988478},
    'miles': {'estimated_diameter_min': 0.0049877647,
     'estimated_diameter_max': 0.0111529809},
    'feet': {'estimated_diameter_min': 26.3354065935,
     'estimated_diameter_max': 58.8877593581}},
   'is_potentially_hazardous_asteroid': False,
   'close_approach_data': [{'close_approach_date': '2025-01-12',
     'close_approach_date_full': '2025-Jan-12 08:32',
     'epoch_date_close_approach': 1736670720000,
     'relative_velocity': {'kilometers_per_second': '10.681

Creo la estructura para el dataframe con los datos de las columnas en español para que sean comprensibles y con los elementos que me interesan que aparezcan.

In [ ]:
# Extraigo el diccionario del objeto data que contiene la información de las fechas indicadas en la petición a la API, el diccionario de tal manera que las claves
# son las fechas y los valores son listas de asteriodes que se acercarán en esas fechas

near_earth_objects = data['near_earth_objects']

# Inicializo una lista vacía donde se almacenarán los datos en formato diccionario, para posteriormente crear el dataframe

all_neo_data = []

# Se realiza un bucle sobre las fechas, se itera sobre cada clave o fecha y se hace un bucle sobre los asteriodes
for date in near_earth_objects:
    for neo in near_earth_objects[date]:
        neo_data = {
            'id': neo['id'],
            'nombre': neo['name'],
            'magnitud_absoluta_h': neo['absolute_magnitude_h'],
            'diametro_min_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_min'],
            'diametro_max_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_max'],
            'potencialmente_peligroso': neo['is_potentially_hazardous_asteroid'],
            'fecha_aproximacion_cercana': neo['close_approach_data'][0]['close_approach_date'],
            'velocidad_relativa_km_per_seg': neo['close_approach_data'][0]['relative_velocity']['kilometers_per_second'],
            'distancia_fallo_km': neo['close_approach_data'][0]['miss_distance']['kilometers'],
            'cuerpo_que_orbita': neo['close_approach_data'][0]['orbiting_body']
        }
        all_neo_data.append(neo_data) # Se crea el diccionario (neo_data) donde se almacena la información de cada asteroide en el diccionario vacío (all_neo_data)

# Convierto los datos en un Dataframe e imprimo el resultado

df_neo = pd.DataFrame(all_neo_data)

# Muestro el resultado

df_neo

id             nombre  magnitud_absoluta_h  \
0    3313157          (2006 BA)               27.600   
1    3648878         (2013 TG6)               26.580   
2    3989277        (2020 BN12)               25.900   
3   54105515         (2021 AR5)               25.550   
4   54111266         (2021 CG4)               25.890   
5   54214729         (2021 UO7)               29.550   
6   54321433        (2022 UB13)               29.920   
7   54395027          (2023 UJ)               26.310   
8   54433329         (2024 GD2)               25.550   
9   54513040        (2024 YZ12)               26.279   
10  54514890        (2024 YB26)               26.472   
11  54515423         (2025 AK3)               26.127   
12   2188174   188174 (2002 JC)               17.440   
13   3615463       (2012 UK171)               24.430   
14   3730322       (2015 TL178)               21.260   
15   3742059        (2016 CF31)               21.730   
16   3773744       (2017 FB157)               27.700   
17   3795086         (2017 YQ5)               24.700   
18   3797456         (2018 AN2)               24.800   
19   3836886         (2018 XA4)               29.000   
20  54144373         (2021 JY3)               28.450   
21  54375205          (2023 OA)               26.450   
22  54514248         (2025 AF1)               27.615   
23  54515075         (2025 AV2)               26.197   
24   2497113  497113 (2004 EK1)               22.280   
25   3167343         (2003 WP7)               23.450   
26   3684344       (2014 QK362)               21.660   
27   3702319          (2014 YD)               24.320   
28   3797861         (2018 BM3)               21.920   
29  54111251         (2021 CO3)               25.780   
30  54244187          (2022 CD)               26.110   
31  54327909         (2022 WW2)               25.570   
32  54457674         (2024 MH2)               23.500   
33  54458508         (2024 NA2)               25.200   
34  54509633         (2024 YD5)               26.265   

    diametro_min_estimado_km  diametro_max_estimado_km  \
0                   0.008027                  0.017949   
1                   0.012840                  0.028710   
2                   0.017561                  0.039268   
3                   0.020633                  0.046136   
4                   0.017642                  0.039449   
5                   0.003270                  0.007312   
6                   0.002758                  0.006167   
7                   0.014540                  0.032512   
8                   0.020633                  0.046136   
9                   0.014749                  0.032979   
10                  0.013494                  0.030174   
11                  0.015818                  0.035370   
12                  0.864082                  1.932146   
13                  0.034559                  0.077275   
14                  0.148784                  0.332690   
15                  0.119827                  0.267941   
16                  0.007666                  0.017141   
17                  0.030518                  0.068240   
18                  0.029144                  0.065169   
19                  0.004213                  0.009420   
20                  0.005427                  0.012135   
21                  0.013632                  0.030482   
22                  0.007972                  0.017825   
23                  0.015316                  0.034248   
24                  0.093015                  0.207989   
25                  0.054269                  0.121350   
26                  0.123753                  0.276720   
27                  0.036354                  0.081291   
28                  0.109788                  0.245493   
29                  0.018559                  0.041499   
30                  0.015942                  0.035648   
31                  0.020443                  0.045713   
32                  0.053034                  0.118588   
33                  0

¿Qué significan estos valores?

* id: Un número único de identificación asignado al asteroide.
* nombre: El nombre oficial del asteroide, seguido de su designación provisional entre paréntesis.
* magnitud_absoluta: Indica el brillo intrínseco del asteroide. Cuanto menor sea el número, más brillante es.
* diametro_min_estimado_km y diametro_max_estimado_km: El rango estimado del diámetro del asteroide en kilómetros.
* potencialmente_peligroso: Un valor booleano (True o False) que indica si el asteroide se considera potencialmente peligroso para la Tierra.
* fecha_aproximacion_cercana: La fecha estimada de la máxima aproximación del asteroide a la Tierra.
* velocidad_relativa_km_per_seg: La velocidad relativa del asteroide con respecto a la Tierra en kilómetros por segundo.
* distancia_fallo_km: La distancia mínima a la que se espera que el asteroide pase de la Tierra, medida en kilómetros.
* cuerpo_que_orbita: El cuerpo celeste alrededor del cual orbita el asteroide (en este caso, la Tierra, pero puede ser otro como la Luna).

Ahora quiero saber de toda esta lista cuales son los potencialmente peligrosos.

In [6]:
cantidad_peligrosos = df_neo['potencialmente_peligroso'].sum()

print(f'Cantidad de objetos potencialmente peligrosos: {cantidad_peligrosos}')

Cantidad de objetos potencialmente peligrosos: 2


Quiero saber cuáles son estos objetos y el detalle de los datos que aparecen

In [7]:
objetos_peligrosos = df_neo['potencialmente_peligroso'] == True

df_neo[objetos_peligrosos]

id        nombre  magnitud_absoluta_h  diametro_min_estimado_km  \
14  3730322  (2015 TL178)                21.26                  0.148784   
28  3797861    (2018 BM3)                21.92                  0.109788   

    diametro_max_estimado_km  potencialmente_peligroso  \
14                  0.332690                      True   
28                  0.245493                      True   

   fecha_aproximacion_cercana velocidad_relativa_km_per_seg  \
14                 2025-01-11                 17.5611867728   
28                 2025-01-10                 22.8423958342   

    distancia_fallo_km cuerpo_que_orbita  
14  49699887.737077757             Earth  
28  21814635.062783591             Earth

Interpretacion de los valores de los objetos considerados peligrosos.

## Asteroide 143730322 (2015 TL178)

* Tamaño: Pequeño, con un diámetro estimado entre 148 y 333 metros.
* Peligrosidad: Se considera potencialmente peligroso.
* Aproximación: Se acercará a la Tierra el 11 de enero de 2025.
* Velocidad: Viajará a una velocidad relativa de aproximadamente 17.56 kilómetros por segundo.
* Distancia: Pasará a una distancia de aproximadamente 49,70 millones de kilómetros de la Tierra.
* Cuerpo que orbita: Tierra.


## Asteroide 283797861 (2018 BM3)

* Tamaño: También es pequeño, con un diámetro estimado entre 109 y 245 metros.
* Peligrosidad: Se considera potencialmente peligroso.
* Aproximación: Se acercará a la Tierra el 10 de enero de 2025, un día antes que el anterior.
* Velocidad: Viajará a una velocidad relativa ligeramente superior, de aproximadamente 22.84 kilómetros por segundo.
* Distancia: Pasará a una distancia de aproximadamente 21,8 millones de kilómetros.
* Cuerpo que orbita: Tierra.

Ambos asteroides están clasificados como potencialmente peligrosos por su tamaño y proximidad. Sin embargo, (2018 BM3) pasa más cerca de la Tierra y viaja a una velocidad mayor, lo que podría justificar un monitoreo más atento.

Pero haciendo una comparación de esa distancia podemos compararla en relación con la Luna, cuya distancia con la Tierra es de 384.400 Kms. Lo que nos indica que en el caso del objeto 2018 BM3 su distancia es de 56,74 veces la distancia de la Tierra a la Luna.(21814635.06/384400 = 56,74 )

Esto significa que el asteroide está relativamente lejos en términos cósmicos, pero sigue siendo lo suficientemente cercano como para ser clasificado como un objeto cercano a la Tierra (NEO).